### PR TITLE
Validate a boolean parameter strictly in annotated services

### LINKED ISSUE #2767: An empty string (or an invalid value) should not be injected as `false` in annotated services
Currently, the query string `?foo=` will inject false in the following service method:

```java
@Get("/")
public String serve(@Param Optional<Boolean> foo) {
    return "foo: " + foo;
}
```

We should probably inject `Optional.empty()` when the query parameter is not a valid boolean value.

/cc @KarboniteKream

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

Given the following annotated service:

    public class MyService {
        @Get("/get")
        public String get(@Param boolean value) {
            return String.valueOf(value);
        }
    }

`GET /get?value=` or `GET /get?value=badvalue` injects `false` into
the parameter `value`, which is unclear to many users.

Modifications:

- Use a custom function for parsing a boolean parameter instead of
  `Boolean.valueOf()`.

Result:

- Less surprising behavior
  - `true`, `false`, `0` and `1` are supported.
- Fixes #2767

### CODE REVIEW

**Comment by heowc** [comment]
Very nice~ :sunglasses:

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTypeUtil.java [hunk 1]
```diff
@@ -115,5 +116,16 @@
                 "Can't convert '" + str + "' to type '" + clazz.getSimpleName() + "'.");
     }
 
+    private static Boolean parseBoolean(String s) {
+        switch (Ascii.toLowerCase(s)) {
+            case "true":
+                return true;
+            case "false":
+                return false;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTypeUtil.java:124]: Couldn't we take other common boolean expressions such as 
- `1` and `0`
- `yes` and `no`
https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/core/convert/support/StringToBooleanConverter.java#L38-L46

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTypeUtil.java:124]: How about `on` and `off` then?

**Comment by trustin** [comment]
I removed `yes`, `no`, `on` and `off` because similarity between `on` and `no` may lead an unexpected consequences, just in case Armeria is used in a nuclear plant. :wink:

**Comment by trustin** [comment]
Thanks for reviewing, folks.

**Pull request merged**